packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -76,75 +83,52 @@ export default util.createRule<Options, MessageId>({
           return;
         }
 
-        for (const expr of node.expressions) {
-          const type = getNodeType(expr);
-          if (!isAllowedType(type)) {
+        for (const expression of node.expressions) {
+          if (
+            !isUnderlyingExpressionTypeConfirmingTo(
+              expression,
+              isUnderlyingTypePrimitive,
+            )
+          ) {
             context.report({
-              node: expr,
+              node: expression,
               messageId: 'invalidType',
             });
           }
         }
       },
     };
 
-    /**
-     * Helper function to get base type of node
-     * @param node the node to be evaluated.
-     */
-    function getNodeType(node: TSESTree.Expression): BaseType[] {
-      const tsNode = service.esTreeNodeToTSNodeMap.get(node);
-      const type = typeChecker.getTypeAtLocation(tsNode);
+    function isUnderlyingExpressionTypeConfirmingTo(
+      expression: TSESTree.Expression,
+      predicate: (underlyingType: ts.Type) => boolean,
+    ): boolean {
+      const expressionType = getExpressionNodeType(expression);
 
-      return getBaseType(type);
-    }
+      return rec(
+        // "Extracts" generic constraint, indexed access and conditional types:
+        typeChecker.getBaseConstraintOfType(expressionType) ?? expressionType,
```

Comment:
We've got a utility function for this!
```suggestion
        util.getConstrainedTypeAtLocation(typeChecker, expression),
```